File: code_to_optimize/java/src/main/java/com/example/StringUtils.java
package com.example;

import java.util.ArrayList;
import java.util.List;

/**
 * String utility functions.
 */
public class StringUtils {

    /**
     * Reverse a string character by character.
     *
     * @param s String to reverse
     * @return Reversed string
     */
    public static String reverseString(String s) {
        if (s == null || s.isEmpty()) {
            return s;
        }

        String result = "";
        for (int i = s.length() - 1; i >= 0; i--) {
            result = result + s.charAt(i);
        }
        return result;
    }

    /**
     * Check if a string is a palindrome.
     *
     * @param s String to check
     * @return true if s is a palindrome
     */
    public static boolean isPalindrome(String s) {
        if (s == null || s.isEmpty()) {
            return true;
        }

        String reversed = reverseString(s);
        return s.equals(reversed);
    }

    /**
     * Count the number of words in a string.
     *
     * @param s String to count words in
     * @return Number of words
     */
    public static int countWords(String s) {
        if (s == null || s.trim().isEmpty()) {
            return 0;
        }

        String[] words = s.trim().split("\\s+");
        return words.length;
    }

    /**
     * Capitalize the first letter of each word.
     *
     * @param s String to capitalize
     * @return String with each word capitalized
     */
    public static String capitalizeWords(String s) {
        if (s == null || s.isEmpty()) {
            return s;
        }

        String[] words = s.split(" ");
        String result = "";

        for (int i = 0; i < words.length; i++) {
            if (words[i].length() > 0) {
                String capitalized = words[i].substring(0, 1).toUpperCase()
                        + words[i].substring(1).toLowerCase();
                result = result + capitalized;
            }
            if (i < words.length - 1) {
                result = result + " ";
            }
        }

        return result;
    }

    /**
     * Count occurrences of a substring in a string.
     *
     * @param s String to search in
     * @param sub Substring to count
     * @return Number of occurrences
     */
    public static int countOccurrences(String s, String sub) {
        if (s == null || sub == null || sub.isEmpty()) {
            return 0;
        }

        int count = 0;
        int index = 0;

        while ((index = s.indexOf(sub, index)) != -1) {
            count++;
            index = index + 1;
        }

        return count;
    }

    /**
     * Remove all whitespace from a string.
     *
     * @param s String to process
     * @return String without whitespace
     */
    public static String removeWhitespace(String s) {
        if (s == null || s.isEmpty()) {
            return s;
        }

        String result = "";
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (!Character.isWhitespace(c)) {
                result = result + c;
            }
        }
        return result;
    }

    /**
     * Find all indices where a character appears in a string.
     *
     * @param s String to search
     * @param c Character to find
     * @return List of indices where character appears
     */
    public static List<Integer> findAllIndices(String s, char c) {
        List<Integer> indices = new ArrayList<>();

        if (s == null || s.isEmpty()) {
            return indices;
        }

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == c) {
                indices.add(i);
            }
        }

        return indices;
    }

    /**
     * Check if a string contains only digits.
     *
     * @param s String to check
     * @return true if string contains only digits
     */
    public static boolean isNumeric(String s) {
        if (s == null || s.isEmpty()) {
            return false;
        }

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c < '0' || c > '9') {
                return false;
            }
        }
        return true;
    }

    /**
     * Repeat a string n times.
     *
     * @param s String to repeat
     * @param n Number of times to repeat
     * @return Repeated string
     */
    public static String repeat(String s, int n) {
        if (s == null || n <= 0) {
            return "";
        }

        String result = "";
        for (int i = 0; i < n; i++) {
            result = result + s;
        }
        return result;
    }

    /**
     * Truncate a string to a maximum length with ellipsis.
     *
     * @param s String to truncate
     * @param maxLength Maximum length (including ellipsis)
     * @return Truncated string
     */
    public static String truncate(String s, int maxLength) {
        if (s == null || maxLength <= 0) {
            return "";
        }

        if (s.length() <= maxLength) {
            return s;
        }

        if (maxLength <= 3) {
            return s.substring(0, maxLength);
        }

        return s.substring(0, maxLength - 3) + "...";
    }

    /**
     * Convert a string to title case.
     *
     * @param s String to convert
     * @return Title case string
     */
    public static String toTitleCase(String s) {
        if (s == null || s.isEmpty()) {
            return s;
        }

        return s.substring(0, 1).toUpperCase() + s.substring(1).toLowerCase();
    }
}

File: code_to_optimize/go/stringutils.go
package sample

import "strings"

func ReverseString(s string) string {
	result := ""
	for i := len(s) - 1; i >= 0; i-- {
		result = result + string(s[i])
	}
	return result
}

func IsPalindrome(s string) bool {
	reversed := ReverseString(s)
	return s == reversed
}

func CountWords(s string) int {
	trimmed := strings.TrimSpace(s)
	if trimmed == "" {
		return 0
	}
	return len(strings.Fields(trimmed))
}

func CapitalizeWords(s string) string {
	if s == "" {
		return s
	}

	words := strings.Split(s, " ")
	result := ""

	for i, word := range words {
		if len(word) > 0 {
			capitalized := strings.ToUpper(word[:1]) + strings.ToLower(word[1:])
			result = result + capitalized
		}
		if i < len(words)-1 {
			result = result + " "
		}
	}
	return result
}

func CountOccurrences(s, sub string) int {
	if sub == "" {
		return 0
	}

	count := 0
	index := 0
	for {
		pos := strings.Index(s[index:], sub)
		if pos == -1 {
			break
		}
		count++
		index = index + pos + 1
	}
	return count
}

func RemoveWhitespace(s string) string {
	result := ""
	for _, c := range s {
		if c != ' ' && c != '\t' && c != '\n' && c != '\r' {
			result = result + string(c)
		}
	}
	return result
}

func FindAllIndices(s string, c byte) []int {
	var indices []int
	for i := 0; i < len(s); i++ {
		if s[i] == c {
			indices = append(indices, i)
		}
	}
	return indices
}

func IsNumeric(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

func Repeat(s string, n int) string {
	if n <= 0 {
		return ""
	}
	result := ""
	for i := 0; i < n; i++ {
		result = result + s
	}
	return result
}

func Truncate(s string, maxLen int) string {
	if maxLen <= 0 {
		return ""
	}
	if len(s) <= maxLen {
		return s
	}
	if maxLen <= 3 {
		return s[:maxLen]
	}
	return s[:maxLen-3] + "..."
}

func ToTitleCase(s string) string {
	if s == "" {
		return s
	}
	return strings.ToUpper(s[:1]) + strings.ToLower(s[1:])
}
